### PR TITLE
[Core] Fix sorting of `Dictionary` keys

### LINKED ISSUE #97522: Animation's data in *.tscn file always change keys order in Dictionary
### Tested versions

- Reproducible in v4.4.dev2.official [97ef3c837]

### System information

Windows 11, Vulkan API 1.3.205 - Forward+ - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce RTX 3060 Ti

### Issue description

When save (without any modification) a scene which contains AnimationPlayers, .tscn file always change for the StringName keys of Dictionary has changed. It's not friendly to version control.

屏幕截图 2024-09-27 112244

### Steps to reproduce

create a scene contains a AnimationPlayer, and make some animation. Save the scene and open the saved tscn file with other txt editor and remember the data( dictionary )'s keys order. Then back to the engine and save the scene again, then recheck the data's keys order.

### Minimal reproduction project (MRP)

NA

### PR DESCRIPTION
**Pull request by AThousandShips**
`StringName` keys were sorted as `StringName` which is unstable.

* Fixes: https://github.com/godotengine/godot/issues/97522

### CODE REVIEW

**Comment by AThousandShips** [comment]
Considering an alternative solution which allows fetching keys with `StringName` converted to `String` to improve behavior, will push as a separate commit soon when tested

Added a few cases I could find of this, will remove any that aren't relevant (the code in `PackedDataContainer` isn't affected as it sorts differently)

File: core/variant/dictionary.h [hunk 1]
```diff
@@ -46,7 +46,7 @@ class Dictionary {
 	void _unref() const;
 
 public:
-	void get_key_list(List<Variant> *p_keys) const;
+	void get_key_list(List<Variant> *p_keys, bool p_string_name_to_string = false) const;
```
Comment (AThousandShips) [inline core/variant/dictionary.h:49]: New alternative solution which fetches keys as `String` when `StringName` was found, replacing the custom sort (doesn't solve the visual shader case though)

File: editor/plugins/visual_shader_editor_plugin.cpp [hunk 2]
```diff
@@ -2129,7 +2129,7 @@ void VisualShaderEditor::_update_nodes() {
 	}
 
 	Array keys = added.keys();
-	keys.sort();
+	keys.sort_custom(callable_mp_static(&StringLikeVariantOrder::compare));
```
Comment (AThousandShips) [inline editor/plugins/visual_shader_editor_plugin.cpp:2132]: If we go with the second solution I'd replace the above method with the list version here

File: core/variant/variant_parser.cpp [hunk 3]
```diff
@@ -2245,7 +2245,7 @@ Error VariantWriter::write(const Variant &p_variant, StoreStringFunc p_store_str
 			} else {
 				List<Variant> keys;
 				dict.get_key_list(&keys);
```
Comment (AThousandShips) [inline core/variant/variant_parser.cpp:2247]: Forgot to update this one with the alternative fix, will push once CI is done

File: servers/rendering_server.cpp [hunk 4]
```diff
@@ -1321,7 +1321,7 @@ Error RenderingServer::mesh_create_surface_data_from_arrays(SurfaceData *r_surfa
 	if (index_array_len) {
 		List<Variant> keys;
 		p_lods.get_key_list(&keys);
-		keys.sort(); // otherwise lod levels may get skipped
+		keys.sort_custom<StringLikeVariantOrder>(); // otherwise lod levels may get skipped
```
Comment (AThousandShips) [inline servers/rendering_server.cpp:1324]:
```suggestion
		keys.sort(); // otherwise lod levels may get skipped
```
Realized this one is just using `float` so will restore

File: core/variant/variant.h [hunk 5]
```diff
@@ -836,6 +836,19 @@ struct StringLikeVariantComparator {
 	static bool compare(const Variant &p_lhs, const Variant &p_rhs);
 };
 
+struct StringLikeVariantOrder {
```
Comment (dalexeev) [inline core/variant/variant.h:839]: Given that there is already `StringLikeVariantComparator`, I have a minor question. As far as I know, there are three main types of comparators:

1. Equality (`a == b`). Not suitable for sorting.
2. Strict or non-strict comparison (`a < b` or `a <= b`).
3. Three-way comparison (`a <=> b` in C++20). Not sure if we use something like this.

It would be nice to make sure that we use consistent naming for these types. But that's not directly relevant to the issue, just a note.

<details>
<summary><code>struct \w+(Compare|Comparator|Sort|Sorter|Order)\b</code> matches</summary>

```
core/config/project_settings.cpp: 1
381:  1: struct _VCSort {
core/debugger/local_debugger.cpp: 1
 37:  2: 	struct ProfileInfoSort {
core/math/convex_hull.cpp: 1
1573:  1: struct PointComparator {
core/string/string_name.h: 1
167:  2: 	struct AlphCompare {
core/string/ustring.h: 3
515:  1: struct NoCaseComparator {
521:  1: struct NaturalNoCaseComparator {
527:  1: struct FileNoCaseComparator {
core/templates/list.h: 1
705:  2: 	struct AuxiliaryComparator {
core/templates/pair.h: 2
 63:  1: struct PairSort {
108:  1: struct KeyValueSort {
core/templates/sort_array.h: 1
 44:  1: struct _DefaultComparator {
core/variant/array.cpp: 1
681:  1: struct _ArrayVariantSort {
core/variant/callable.h: 1
203:  1: struct CallableComparator {
core/variant/variant.h: 2
831:  1: struct VariantComparator {
835:  1: struct StringLikeVariantComparator {
drivers/gles3/rasterizer_scene_gles3.h: 1
155:  2: 	struct InstanceSort {
drivers/gles3/storage/particles_storage.h: 1
 71:  2: 	struct ParticlesViewSort {
editor/animation_bezier_editor.h: 1
172:  2: 	struct PairCompare {
editor/connections_dialog.cpp: 1
919:  1: struct _ConnectionsDockMethodInfoSort {
editor/debugger/script_editor_debugger.h: 1
173:  2: 	struct ThreadSort {
editor/doc_tools.cpp: 3
 80:  1: struct ConstructorCompare {
 98:  1: struct OperatorCompare {
114:  1: struct MethodCompare {
editor/editor_command_palette.h: 2
 62:  2: 	struct CommandEntryComparator {
 68:  2: 	struct CommandHistoryComparator {
editor/editor_file_system.h: 1
231:  2: 	struct DirectoryComparator {
editor/editor_help.h: 1
204:  2: 	struct PropertyCompare {
editor/editor_quick_open.h: 1
 58:  2: 	struct EntryComparator {
editor/editor_settings.cpp: 1
238:  1: struct _EVCSort {
editor/file_info.h: 2
 60:  1: struct FileInfoTypeComparator {
 66:  1: struct FileInfoModifiedTimeComparator {
editor/groups_editor.cpp: 1
 66:  1: struct _GroupInfoComparator {
editor/plugins/mesh_instance_3d_editor_plugin.cpp: 1
359:  1: struct MeshInstance3DEditorEdgeSort {
editor/plugins/node_3d_editor_plugin.cpp: 2
9291:  1: struct _GizmoPluginPriorityComparator {
9300:  1: struct _GizmoPluginNameComparator {
editor/plugins/node_3d_editor_plugin.h: 2
 74:  2: 	struct Axis2DCompare {
420:  2: 	struct ShortcutCheckSetComparator {
editor/plugins/text_shader_editor.h: 1
 61:  2: 	struct WarningsComparator {
editor/plugins/tiles/tiles_editor_plugin.h: 1
 64:  2: 	struct SourceNameComparator {
editor/plugins/visual_shader_editor_plugin.h: 1
403:  2: 	struct _OptionComparator {
editor/project_manager/project_list.cpp: 1
322:  1: struct ProjectListComparator {
modules/gdscript/gdscript.cpp: 2
280:  1: struct _GDScriptMemberSort {
2496:  1: struct GDScriptDepSort {
modules/gridmap/editor/grid_map_editor_plugin.cpp: 1
791:  1: struct _CGMEItemSort {
modules/lightmapper_rd/lightmapper_rd.h: 1
206:  2: 	struct TriangleSort {
modules/mono/csharp_script.cpp: 1
548:  1: struct CSharpScriptDepSort {
modules/text_server_adv/text_server_adv.h: 1
694:  2: 	struct GlyphCompare { // For line breaking reordering.
scene/2d/tile_map_layer.h: 2
164:  1: struct CellDataYSortedXReversedComparator {
195:  2: 	struct CoordsWorldComparator {
scene/3d/gpu_particles_collision_3d.h: 1
142:  2: 	struct FaceSort {
scene/3d/voxelizer.h: 1
 64:  2: 	struct CellSort {
scene/gui/code_edit.h: 1
515:  1: struct CodeCompletionOptionCompare {
scene/gui/control.h: 1
156:  2: 	struct CComparator {
scene/gui/text_edit.cpp: 1
4847:  1: struct _CaretSortComparator {
scene/main/scene_tree.h: 1
106:  2: 	struct ProcessGroupSort {
scene/resources/3d/importer_mesh.h: 1
 66:  3: 		struct LODComparator {
scene/resources/resource_format_text.h: 1
188:  2: 	struct ResourceSort {
scene/resources/surface_tool.h: 1
128:  2: 	struct WeightSort {
servers/debugger/servers_debugger.cpp: 1
197:  2: 	struct ProfileInfoSort {
servers/rendering/renderer_canvas_cull.h: 2
 93:  2: 	struct ItemIndexSort {
 99:  2: 	struct ItemPtrSort {
servers/rendering/renderer_rd/storage_rd/light_storage.h: 2
165:  2: 	struct LightInstanceDepthSort {
312:  2: 	struct ReflectionProbeInstanceSort {
servers/rendering/renderer_rd/storage_rd/texture_storage.h: 1
312:  2: 	struct DecalInstanceSort {
servers/rendering/rendering_device_graph.h: 1
232:  2: 	struct RecordedCommandSort {
servers/rendering/shader_language.h: 1
748:  2: 	struct UniformOrderComparator {
```

</details>

Comment (AThousandShips) [reply core/variant/variant.h:839]: If we were to standardize this I'd change the ones named `Compare` but does equality checks to `Eq` or similar and the ones comparing order to `Compare` (as it seems the most common for those)

**Comment by AThousandShips** [comment]
Added a unit test for the `Dictionary` case, there are no `stringify` tests currently for JSON so left that for now, but can add if desired

File: core/variant/variant.h [hunk 6]
```diff
@@ -836,6 +836,19 @@ struct StringLikeVariantComparator {
...
+	static _ALWAYS_INLINE_ bool compare(const Variant &p_lhs, const Variant &p_rhs) {
```
Comment (AThousandShips) [inline core/variant/variant.h:840]: This static method isn't technically needed with the alternative solution (or when using `List<Variant>` in the visual shader code) but I think it's useful to keep so we can use it when sorting `Array` as it takes a `Callable` and we can use `callable_mp_static`

I will look at adding such static methods to other comparators to aid in using them for such cases

File: core/io/json.cpp [hunk 7]
```diff
@@ -118,7 +118,7 @@ String JSON::_stringify(const Variant &p_var, const String &p_indent, int p_cur_
 			p_markers.insert(d.id());
 
 			List<Variant> keys;
-			d.get_key_list(&keys);
+			d.get_key_list(&keys, true);
```
Comment (rune-scape) [inline core/io/json.cpp:121]: why not use `keys.sort_custom<StringLikeVariantOrder>()` instead of this interface change?

Comment (AThousandShips) [reply core/io/json.cpp:121]: Look above, this is an alternative solution I'm proposing 🙃

File: core/variant/variant.h [hunk 8]
```diff
@@ -836,6 +836,19 @@ struct StringLikeVariantComparator {
...
+		if (p_lhs.is_string() && p_rhs.is_string()) {
+			return p_lhs.operator String() < p_rhs.operator String();
```
Comment (Macksaur) [inline core/variant/variant.h:842]: Allocating memory, filling it, then doing a comparison on the copy before dropping it immediately on the floor feels very slapdash. The memory should already be meaningful and available to use for such a comparison.

Perhaps this can be reused?:
https://github.com/godotengine/godot/blob/f032af74536b317b23c7fca3bc7318ced5537344/core/string/ustring.h#L534

Comment (AThousandShips) [reply core/variant/variant.h:842]: It doesn't allocate memory except when the `StringName` is created via a c-strinf, which is rare comparatively

Comment (Macksaur) [reply core/variant/variant.h:842]: I'll take it, thank you!

Comment (AThousandShips) [reply core/variant/variant.h:842]: But will keep in mind for a future helper allowing comparison with both `String` and `StringName` generally using this correctly!

**Comment by AThousandShips** [comment]
Removed the alternative fix and will look at it for the future, to get this out there soon

Was looking at the new sort method of `Dictionary` which I had forgotten to look at after it was added for this purpose, but it doesn't suite this without some changes as it suffers from the same ordering limitations, namely preserving compatibility for `Variant` ordering in general

So for context, this solves that `OP_LESS` of `Variant` sorts by type, so all `String` keys are sorted separately and `StringName` keys are sorted separately, further `StringName` keys are sorted by pointer for performance, but that's unstable

This was introduced by changes that removed the conversion from `StringName` to `String` on all keys in:
* #70096 

Tl;Dr;
This aims to solve sorting of `Dictionary` keys in serialisation in a compatibility preserving way while not impacting performance in general

**Comment by fire** [comment]
Does this fix the animation tscn sorting issue?

**Comment by AThousandShips** [comment]
Yes it should, will confirm fully today if no one else has

**Review by KoBeWi — APPROVED** [review]

**Comment by Repiteo** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank *you*!
